### PR TITLE
feat(api): expose various lsp state calls and refactor api slightly

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**

https://app.clubhouse.io/uma-project/story/1714/expose-lsp-state-through-api

**Summary**

This exposes LSP state queried through ingestion services to the express api.

**Details**
There was some refactoring done to how the API works, but it should remain backward compatible. Specific contract calls to the api are now separated by "channels", so the path prepends `/emp/` or `/lsp/` to the action for instance `localhost:8282/emp/actions` - to list all actions on the channel.  The root channel `/` and `/emp` are identical for backward compatibility to existing clients, but this is deprecated and will probably change in the future.

Several new API calls were added specifically for the LSP which mirror many of the EMP calls. The README summarizes these.

There is a new [postman link ](https://www.getpostman.com/collections/9ce0fe57a483b5163fbf) which can be imported with the updated calls.

The action-client example has also been updated to show how to use the new channel system.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**
https://app.clubhouse.io/uma-project/story/1714/expose-lsp-state-through-api

### CODE REVIEW

File: packages/api/src/apps/api/README.md [hunk 1]
```diff
@@ -63,128 +63,170 @@ This is a non exhaustive list of actions, for the full list see [action source](
...
 
 Returns a list of all actions
 
-### echo(...args:Json[]) => args
+### emp/echo(...args:Json[]) => args
 
 Echos back any parameters you send to the server. For testing connectivity and client correctness.
 
-### listEmpAddresses() => string[]
+### emp/listEmpAddresses() => string[]
```
Comment (mrice32) [inline packages/api/src/apps/api/README.md:92]: Does this mean that the user has to figure out which type of contract a particular address is to get information on it? Is there a reason why we can't just offer all calls for any address and have an error message if that call/action isn't available for that action?

While this may be more elegant from a design/implementation perspective, it seems harder for the user or frontend to navigate because there are 1) more functions and 2) they need to have a deeper understanding of the contract that they're querying to actually make a query.

I think an illustrative example is etherscan -- a user shows up with an address and etherscan gives them curated information based on what it knows about that address without the user needing to request that information specifically.

Also, is there a way to ask the API for the type of a contract or is the expectation that the frontend figure that out?

**Review by mrice32 — COMMENTED** [review]
Looks awesome! One high-level comment about the separation of the emp api and the lsp api. We might want to have a discussion offline about this.

File: packages/api/src/apps/api/README.md (hunk 1 above)
Comment (daywiss) [reply packages/api/src/apps/api/README.md:92]: yea i totally get this comment. currently yes, it requires specific knowledge to call the right endpoint.  this is just a first step though towards the feature you are describing. I plan to use the root directory for "global" style calls which would include searching for contracts across types, as well as any kind of stats which combine these or other types together, like tvl. 

I didnt want to do that all in this pr since i wanted to try and maintain backward compatibility for now (root path points to the same calls as previous version), and it makes sense as a separate feature from this.

Comment (mrice32) [reply packages/api/src/apps/api/README.md:92]: If this PR is only to add the capability to build calls from root that can be on any contract type, then why do we need to expose these channels to the user at all?

**Review by mrice32 — COMMENTED** [review]
Annoying that github doesn't let you just leave a single comment as a review 🙃. https://github.com/UMAprotocol/protocol/pull/3245#discussion_r674743236

File: packages/api/src/apps/api/README.md (hunk 1 above)
Comment (daywiss) [reply packages/api/src/apps/api/README.md:92]: hmm good question. This is a higher level design discussion.  I ideally want to unblock FE work and defer this design decision.  This PR will allow FE to start integrating under the assumption they first know if an address is an LSP or EMP.  The only thing needed later in the global scope would be some calls to inform what type of contract a particular address is  and pass the address to the correct contract template. 

There's an alternative design i think, which is maybe what you are referring to, where the global scope contains all the calls for each contract type, and automatically switches based on the address.  Its not my favorite design, but will definitely take input from FE based on what they would prefer.  I think this decision can be deferred somewhat because this will at least  this will unblock some FE development and does not rule out either design.

**Review by mrice32 — APPROVED** [review]
As discussed offline, we may want to spec these api endpoints out ahead of time, but the code LGTM!

File: packages/api/src/services/express-channels.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import { Json, BaseConfig } from "..";
+import { Profile } from "../libs/utils";
+
+interface Config extends BaseConfig {
+  port: number;
+  path?: string;
+}
+// represents the actions service, a single function which maps to different callable functions
+type Actions = (action: string, ...args: Json[]) => Promise<Json>;
+type Channel = [string, Actions];
```
Comment (chrismaree) [inline packages/api/src/services/express-channels.ts:16]: this is cool. I like this ideas for building up dynamic paths

File: packages/api/src/apps/api/README.md (hunk 1 above)
Comment (chrismaree) [reply packages/api/src/apps/api/README.md:92]: I agree that in an ideal world we don't have levels like this. a user should be able to query any address, just like on etherscan, and from this find if it is: a) an LSP, b) a EMP c) an LSP long or short token, d) and EMP synthetic f) even an LSP FNP! there should be no differentiation at the query level. 

but, from a design perspective, it does make sense to separate these into separate logical containers to keep things somewhat modular. I guess it is fine to defer this design discussion until all the functionality is built out, assuming you feel confident that you can easily pull all these end points together through some aggregator end point?

Unblocking the FE is good but not if it means incurring technical/design debt.

**Review by chrismaree — APPROVED** [review]

File: packages/api/src/apps/api/README.md (hunk 1 above)
Comment (daywiss) [reply packages/api/src/apps/api/README.md:92]: > but, from a design perspective, it does make sense to separate these into separate logical containers to keep things somewhat modular. I guess it is fine to defer this design discussion until all the functionality is built out, assuming you feel confident that you can easily pull all these end points together through some aggregator end point?

thanks for discussion @chrismaree , also talked with @mrice32 a bit more about this offline. I think there is some implicit knowledge on my end which ends up making this particular design a bit suspect to others, including previous discussions around wanting to be able to query across contracts. Heres why I think this is appropriate, even if its eventually deprecated in favor of a unified api:

- discussed this design with @Gamaranto 1 on 1 and got his go ahead with this as a first pass, though did not document this publicly and probably should have
- from our discussion he also assured me that its also very easy to adapt his api client to changes in the API, which in my mind means that exposing standalone LSP api now is better than waiting until the more complex combined api, so the LSP logic can be vetted sooner (this is a big assumption and should probably run this again by francesco as it may not be true)
- moving contracts types to specific channels opens up the root channel for unified queries across all contracts, which was a big motivator for this change
- api calls contain little to no business logic, meaning from a development perspective, are almost free, the underlying business logic will not change to make a unified api
- unified api requires more logic in the api layer to allow switching between contract types, which i feel adds another new feature on an already new unvetted feature (LSPs) and makes sense as a separate PR.

Long story short though, the unified API is still on the table and will be discussing this in next infra meeting. The extra work to add contract specific api routes are minor (vs just not exposing them at all), so i see it as adding some flexibility for very little cost.
